spring-cloud-gcp-kms/src/main/java/com/google/cloud/spring/kms/KMSTemplate.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright 2017-2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.spring.kms;
+
+import java.util.Base64;
+
+import com.google.cloud.kms.v1.CryptoKeyName;
+import com.google.cloud.kms.v1.DecryptRequest;
+import com.google.cloud.kms.v1.DecryptResponse;
+import com.google.cloud.kms.v1.EncryptRequest;
+import com.google.cloud.kms.v1.EncryptResponse;
+import com.google.cloud.kms.v1.KeyManagementServiceClient;
+import com.google.cloud.spring.core.GcpProjectIdProvider;
+import com.google.protobuf.ByteString;
+
+/**
+ * Offers convenience methods for performing common operations on KMS including
+ * encrypting and decrypting text.
+ *
+ * @author Emmanouil Gkatziouras
+ */
+public class KMSTemplate implements KMSOperations {
+
+	private final KeyManagementServiceClient client;
+
+	private final GcpProjectIdProvider projectIdProvider;
+
+	public KMSTemplate(
+			KeyManagementServiceClient keyManagementServiceClient,
+			GcpProjectIdProvider projectIdProvider) {
+		this.client = keyManagementServiceClient;
+		this.projectIdProvider = projectIdProvider;
+	}
+
+	public String encrypt(String cryptoKey, String plaintext) {
+		CryptoKeyName cryptoKeyName = KMSPropertyUtils.getCryptoKeyName(cryptoKey, projectIdProvider);
+
+		ByteString plaintextByteString = ByteString.copyFromUtf8(plaintext);
+
+		EncryptRequest request = EncryptRequest.newBuilder()
+				.setName(cryptoKeyName.toString())
+				.setPlaintext(plaintextByteString)
+				.build();
+
+		EncryptResponse response = client.encrypt(request);
+
+		return encodeBase64(response);
```

Comment:
Why do you prefer to do base64 encoding vs. the plain String with `response.getCiphertext().toStringUtf8()`?

Reply:
If I try to do the following there is a failure on decrypting

```
		KeyManagementServiceClient client = KeyManagementServiceClient.create();

		CryptoKeyName cryptoKeyName = CryptoKeyName.newBuilder()
				.setProject("project-id")
				.setLocation("europe-west2")
				.setKeyRing("spring-cloud-gcp")
				.setCryptoKey("key-id")
				.build();

		String encrypted = client.encrypt(cryptoKeyName.toString(), ByteString.copyFromUtf8( "1234")).getCiphertext().toStringUtf8();
		String decrypted = client.decrypt(cryptoKeyName.toString(), ByteString.copyFromUtf8(encrypted)).getPlaintext().toStringUtf8();

		Assert.assertEquals("1234",decrypted);
```

`Caused by: io.grpc.StatusRuntimeException: INVALID_ARGUMENT: Decryption failed: the ciphertext is invalid.`

Also the text is gibberish, something like

```
$ !YЋ��
������^%����������R��fL*7�9- �r��p�];#��v���c&<�w��	��I�R_r���AQ�k�$z�
```

Base64 helped me a lot due to portability (storing to db, passing around) and representation.

Reply:
Hi @dzou & @meltsufin 
I do see Base64 to be a bit opinionated.

In any case I have a branch [here](https://github.com/gkatzioura/spring-cloud-gcp-1/tree/kms-base64-proposal/spring-cloud-gcp-kms/src/main/java/com/google/cloud/spring/kms)
with a proposal on returning back an object, that can give the encryption results either as byte[] or a Base64 string.

```
public class EncryptionResult {

	private final ByteString byteString;

	EncryptionResult(ByteString byteString) {
		this.byteString = byteString;
	}

	public byte[] toByteArray() {
		return byteString.toByteArray();
	}

	public String toBase64() {
		byte[] bytes = byteString.toByteArray();
		byte[] encoded = Base64.getEncoder().encode(bytes);
		return new String(encoded);
	}

}
```

```
public DecryptionResult decrypt(String cryptoKey, byte[] bytes) {
		CryptoKeyName cryptoKeyName = KmsPropertyUtils.getCryptoKeyName(cryptoKey, projectIdProvider);

		ByteString encryptedByteString = ByteString.copyFrom(bytes);
...
		return new DecryptionResult(response.getPlaintext());
	}
```

Let me know if this could be better.